gst/nnstreamer/nnstreamer_subplugin.c

```diff
@@ -66,24 +63,65 @@ static subpluginSearchLogic searchAlgorithm[] = {
   [NNS_SUBPLUGIN_END] = NNS_SEARCH_NO_OP,
 };
 
-/** @brief Public function defined in the header */
-const void *
-get_subplugin (subpluginType type, const char *name)
+/**
+ * @brief Internal function to get sub-plugin data.
+ */
+static subpluginData *
+_get_subplugin_data (subpluginType type, const gchar * name)
 {
-  GHashTable *table;
   subpluginData *spdata = NULL;
-  void *handle;
-
-  g_return_val_if_fail (name, NULL);
 
   G_LOCK (splock);
-
-  if (subplugins[type] == NULL)
+  if (subplugins[type] == NULL) {
     subplugins[type] =
         g_hash_table_new_full (g_str_hash, g_str_equal, g_free,
         _spdata_destroy);
+  } else {
+    spdata = g_hash_table_lookup (subplugins[type], name);
+  }
+  G_UNLOCK (splock);
+
+  return spdata;
+}
+
+/**
+ * @brief Internal function to scan sub-plugin.
+ */
+static subpluginData *
+_search_subplugin (subpluginType type, const gchar * name, const gchar * path)
+{
+  subpluginData *spdata = NULL;
+  void *handle;
 
-  table = subplugins[type];
+  g_return_val_if_fail (name != NULL, NULL);
+  g_return_val_if_fail (path != NULL, NULL);
+
+  dlerror ();
+  handle = dlopen (path, RTLD_NOW);
+  /* If this is a correct subplugin, it will register itself */
+  if (handle == NULL) {
+    g_critical ("Cannot dlopen %s(%s) with error %s.", name, path, dlerror ());
+    return NULL;
+  }
+
+  spdata = _get_subplugin_data (type, name);
+  if (spdata == NULL) {
+    g_critical
+        ("nnstreamer_subplugin of %s(%s) is broken. It does not call register_subplugin with its init function.",
+        name, path);
+    dlclose (handle);
```

Comment:
I think the handle should be closed outside the if statement.

Reply:
No. the handle should be kept alive if its contents (the subplugin) are to be used elsewhere. That's why I'm working on #2049 although I'm experiencing strange behaviors with it.